### PR TITLE
Improve performance for dpnp.diag()

### PR DESCRIPTION
**Pull request by vlad-perevezentsev**
This PR suggests to update `dpnp.diag()` implementation with a reuse of `dpnp.diagonal()` and return a read/write view instead of copy if the input array is an instance of {`dpnp.ndarray`, `usm_ndarray`} for the default case ( `device` ,`usm_type`, `sycl_queue` are `None`).
These changes greatly improve performance compared to the old implementation

image

- [X] Have you provided a meaningful PR description?
- [X] Have you added a test, reproducer or referred to issue with a reproducer?
- [X] Have you tested your changes locally for CPU and GPU devices?
- [X] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: dpnp/dpnp_iface_arraycreation.py [hunk 1]
```diff
@@ -894,79 +899,67 @@ def diag(v, /, k=0, *, device=None, usm_type=None, sycl_queue=None):
...
-    >>> y, y.device, y.usm_type
+    >>> res_cpu = np.diag(x, device="cpu")
+    >>> res_cpu, res_cpu.device, res_cpu.usm_type
     (array([0, 4, 8]), Device(opencl:cpu:0), 'device')
 
-    >>> z = np.diag(x0, usm_type="host")
-    >>> z, z.device, z.usm_type
+    >>> res_host = np.diag(x, usm_type="host")
+    >>> res_host, res_host.device, res_host.usm_type
     (array([0, 4, 8]), Device(level_zero:gpu:0), 'host')
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_arraycreation.py:940]: `See Also` link is missing from `dpnp.eye` description to `diag` function

File: dpnp/dpnp_iface_arraycreation.py [hunk 2]
```diff
@@ -894,79 +899,67 @@ def diag(v, /, k=0, *, device=None, usm_type=None, sycl_queue=None):
...
 
     """
 
     if not isinstance(k, int):
         raise TypeError(f"An integer is required, but got {type(k)}")
 
-    v = dpnp.asarray(v, device=device, usm_type=usm_type, sycl_queue=sycl_queue)
+    v = dpnp.asanyarray(
```
[the remark this answers is not in the document]
Comment (vlad-perevezentsev) [reply dpnp/dpnp_iface_arraycreation.py:946]: done

File: dpnp/dpnp_iface_arraycreation.py (hunk 1 above)
Comment (vlad-perevezentsev) [reply dpnp/dpnp_iface_arraycreation.py:940]: Thank you for noticing that!
I added it.

**Review by antonwolfy — APPROVED** [review]
Thank you @vlad-perevezentsev for implementing so noticeable improvement.

**Pull request merged**